integration/helpers.go

```diff
@@ -1547,15 +1547,76 @@ func startAndWait(process *service.TeleportProcess, expectedEvents []string) ([]
 	return receivedEvents, nil
 }
 
-type proxyServer struct {
+type proxyBasicAuthHandler struct {
+	sync.Mutex
+	lastError error
+	authDB    map[string]string
+	proxyHandler
+}
+
+func (p *proxyBasicAuthHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
The problem I see with this is that I want to access the state of the proxyBasicAuthHandler. If I follow this suggestion then we could do something like:

```go
	ph := &proxyBasicAuthHandler{authDB: map[string]string{validUser: validPass}}
	wrappedHandler1 := ph.withProxyAuth(&proxyHandler{})
	// ...
	wrappedHandler2 := ph.withProxyAuth(&someOtherHandler{})
	// ...
	// ph.lastError refers to last error from which handler?
```

If I changed it so that the withProxyAuth function did not have a method receiver and instead created a new proxyAuthHandler inside the function, then that would work to prevent this subtle mistake but we would lose access to the struct's state.

I do like how this suggestion decouples proxy auth from a specific handler though. How about this:

```go
type proxyAuthorizer struct {
	next http.Handler
	sync.Mutex
	lastError error
	authDB    map[string]string
}

func NewProxyAuthorizer(handler http.Handler, authDB map[string]string) *proxyAuthorizer {
	return &proxyAuthorizer{next: handler, authDB: authDB}
}

func (p *proxyAuthorizer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	auth := r.Header.Get("Proxy-Authorization")
	if auth == "" {
		err := trace.AccessDenied("missing Proxy-Authorization header")
		p.SetError(err)
		trace.WriteError(w, err)
		return
	}
	user, password, ok := parseProxyAuth(auth)
	if !ok {
		err := trace.AccessDenied("bad Proxy-Authorization header")
		p.SetError(err)
		trace.WriteError(w, err)
		return
	}

	if p.isAuthorized(user, password) {
		p.SetError(nil)
		p.next.ServeHTTP(w, r)
		return
	}

	err := trace.AccessDenied("bad credentials")
	p.SetError(err)
	trace.WriteError(w, err)
}
```